### PR TITLE
Trilinos extra

### PR DESCRIPTION
**Pull request by davydden**
- [x] make it work on OS-X, tested with Clang@7.0.2 / GNU Fortran@5.3
- [x] make ti work on Ubuntu, tested with GNU C/C++/Fortran@4.8
- [x]  fix install_name's on Darwin

blocked by:
https://github.com/LLNL/spack/issues/634 (OSX only)

### CODE REVIEW

File: var/spack/repos/builtin/packages/trilinos/package.py [hunk 1]
```diff
@@ -27,27 +27,141 @@ class Trilinos(Package):
     depends_on('matio')
     depends_on('glm')
     depends_on('swig')
+    depends_on('metis')
+    depends_on('suite-sparse')
+    depends_on('tbb')
```
Comment (Rombur) [inline var/spack/repos/builtin/packages/trilinos/package.py]: I don't think that Trilinos supports tbb. The deprecated version of Kokkos did but the new one doesn't.

File: var/spack/repos/builtin/packages/trilinos/package.py [hunk 2]
```diff
@@ -27,27 +27,141 @@ class Trilinos(Package):
...
+                        '-DBoost_INCLUDE_DIRS:PATH=%s' % spec['boost'].prefix.include,
+                        '-DBoost_LIBRARY_DIRS:PATH=%s' % spec['boost'].prefix.lib,
+                        # '-DTrilinos_ENABLE_Fortran=OFF', # FIXME
+                        '-DTrilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON',
+                        '-DTrilinos_ENABLE_CXX11:BOOL=ON',
+                        '-DTrilinos_CXX11_FLAGS=-std=c++11',
+                        '-DTPL_ENABLE_Netcdf:BOOL=ON',
+                        '-DTPL_ENABLE_HYPRE:BOOL=ON',
+                        '-DTPL_ENABLE_HDF5:BOOL=ON',
+                        '-DTPL_ENABLE_TBB:BOOL=ON',
```
Comment (Rombur) [inline var/spack/repos/builtin/packages/trilinos/package.py]: I don't think this does anything anymore.

File: var/spack/repos/builtin/packages/trilinos/package.py (hunk 1 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/trilinos/package.py]: oh, interesting. Thanks. Will remove it...

**Comment by tgamblin** [comment]
does it make sense to make the (rather extensive) extra stuff into variants?  I may end up doing this anyway to support other Trilinos use cases, so it would be good if `deal.ii` and other dependents were explicit about what they need.

**Comment by davydden** [comment]
@tgamblin agree, will do so after I make it compile. So far have some issues with MUMPS, playing around to make MUMPS build dynamic libs (as that did work for me previously).

**Comment by tgamblin** [comment]
@davydden: sounds good to me.  Trillions is huge so no rush.  I kind of wish it could be broken up into smaller pieces, but it's not distributed that way.  It almost needs an intra-variant dep system to build properly.

**Comment by davydden** [comment]
blocked by https://github.com/LLNL/spack/pull/626

**Comment by davydden** [comment]
@tgamblin @alalazo @Rombur:  with an exception of blocking Mumps PR https://github.com/LLNL/spack/pull/626, i think I am done with extending Trilinos. Any comments to this PR? p.s. i will stash when everyone's happy.

File: var/spack/repos/builtin/packages/trilinos/package.py [hunk 3]
```diff
@@ -27,28 +42,198 @@ class Trilinos(Package):
...
+                        '-DLAPACK_LIBRARY_NAMES=lapack',
+                        '-DLAPACK_LIBRARY_DIRS=%s' % spec['lapack'].prefix,
+                        '-DTPL_ENABLE_Boost:BOOL=ON',
+                        '-DBoost_INCLUDE_DIRS:PATH=%s' % spec['boost'].prefix.include,
+                        '-DBoost_LIBRARY_DIRS:PATH=%s' % spec['boost'].prefix.lib,
+                        '-DTrilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON',
+                        '-DTrilinos_ENABLE_CXX11:BOOL=ON',
+                        '-DTPL_ENABLE_Netcdf:BOOL=ON',
+                        '-DTPL_ENABLE_HYPRE:BOOL=ON',
+                        '-DTPL_ENABLE_HDF5:BOOL=ON'
```
Comment (Rombur) [inline var/spack/repos/builtin/packages/trilinos/package.py]: I think this will crash if I do ```spack install trilinos ~hdf5```

**Comment by Rombur** [comment]
It looks good to me except for ```HDF5```

**Comment by davydden** [comment]
@Rombur thanks Bruno!. It was also the case for `hypre`. If you would like to try it yourself, i keep a branch where I merge all the opened PR's https://github.com/davydden/spack/tree/dealii_stable . Should work OOB with GNU compilers on Ubuntu.

**Comment by davydden** [comment]
@tgamblin let me know when you will have a look at Trilinos PR. This should be ok to merge, but I need to stash things before (there are 42! commits).

**Comment by tgamblin** [comment]
@davydden: I was able to get this built on our systems just fine, but it needed a minor modification to force `suite-sparse` to use spack compilers.  Thanks!

**Comment by davydden** [comment]
@tgamblin great! Glad to hear that!

**Pull request merged**